### PR TITLE
Minimize glActiveTexture calls

### PR DESCRIPTION
**Pull request by snagy**
**Description**

When running spector.js captures of a complex scene, I noticed a number of calls to glActiveTexture that were followed immediately by a separate call to glActiveTexture.  This change checks to see if a texture will be bound (or uploaded) before calling ActiveTexture, improving performance by removing these useless gl calls.

*This contribution is funded by [Meta](https://meta.com)*

### CODE REVIEW

**Comment by Mugen87** [comment]
This is an interesting PR, thanks! Do you think it's possible to reproduce the redundant gl calls in a small live example? I've prepared a fiddle with a simple textured mesh. Do you think you can modify/enhance it so we have something to test?

https://jsfiddle.net/w290bLj3/

File: src/renderers/webgl/WebGLTextures.js [hunk 1]
```diff
@@ -1040,11 +1037,12 @@ function WebGLTextures( _gl, extensions, state, properties, capabilities, utils,
 		const forceUpload = initTexture( textureProperties, texture );
 		const source = texture.source;
 
-		state.activeTexture( _gl.TEXTURE0 + slot );
-		state.bindTexture( _gl.TEXTURE_CUBE_MAP, textureProperties.__webglTexture );
+		state.bindTextureToSlot( _gl.TEXTURE0 + slot, _gl.TEXTURE_CUBE_MAP, textureProperties.__webglTexture );
 
 		if ( source.version !== source.__currentVersion || forceUpload === true ) {
 
+			state.activeTexture( _gl.TEXTURE0 + slot );
```
Comment (Mugen87) [inline src/renderers/webgl/WebGLTextures.js:1044]: Why is it necessary to call `activeTexture()` here when `bindTextureToSlot()` was executed right before?

**Comment by snagy** [comment]
@Mugen87 The CSM example shows the issue pretty well - we're using CSM so that's where I initially saw it.

Without this change it calls ActiveTexture 4 times (for the 4 shadow sections) when it changes material:
Screen Shot 2022-08-15 at 10 27 00 AM

And after applying this change it doesn't call it at all because the textures were already bound to that slot:
Screen Shot 2022-08-15 at 10 32 18 AM

Because this example only uses 2 materials, you don't see a ton of savings, but on our scenes with 80+ material changes the gl calls add up to a significant chunk of cpu time doing nothing.

File: src/renderers/webgl/WebGLTextures.js (hunk 1 above)
Comment (snagy) [reply src/renderers/webgl/WebGLTextures.js:1044]: It's possible that the version was bumped for an already bound texture.  If the texture was already bound to a slot,  BindTextureToSlot exits without calling activeTexture.  Fortunately activeTexture on the three.js side already caches which slot is active, so if we've bound the texture and then drop into here, we simply check the cache and exit out without duplicating the gl call.

Comment (snagy) [reply src/renderers/webgl/WebGLTextures.js:1044]: Also, it's not theoretical: we had a case where we were preloading bone textures and they weren't updating properly before I added this.

**Comment by mrdoob** [comment]
Nice!

Is is worth having both `state.bindTexture` and `state.bindTextureToSlot` though?

Would it make sense to update all the current calls to `state.bindTexture` to `state.bindTextureToSlot` and then delete `state.bindTexture` and rename `state.bindTextureToSlot` to `state.bindTexture`?

**Comment by snagy** [comment]
> Nice!
> 
> Is is worth having both `state.bindTexture` and `state.bindTextureToSlot` though?
> 
> Would it make sense to update all the current calls to `state.bindTexture` to `state.bindTextureToSlot` and then delete `state.bindTexture` and rename `state.bindTextureToSlot` to `state.bindTexture`?

We could definitely do that.  The remaining usage of bindTexture is all around render target stuff (generating mipmaps, setting up framebuffer textures) that doesn't currently do any slot management.  The remaining question is whether calling it with slot == undefined should set the slot to currentTextureSlot, or to the last slot.  currentTextureSlot would minimize activeTexture calls and match the existing behavior the closest.  Using the last slot could potentially keep from rebinding over textures, but that's more content dependent.

**Comment by mrdoob** [comment]
So could it be this instead and make it work for both?

```js
function bindTexture( webglType, webglTexture, webglSlot ) {
```

**Comment by mrdoob** [comment]
Much nicer!

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**